components/supervisor/pkg/supervisor/tasks.go

```diff
@@ -0,0 +1,412 @@
+		legacyPrebuildLogFileName := "/workspace/.prebuild-log-" + task.Id
+		printlogs := "[ -r " + legacyPrebuildLogFileName + " ] && cat " + legacyPrebuildLogFileName + "; [ -r " + prebuildLogFileName + " ] && cat " + prebuildLogFileName + "; true"
+		return []*string{task.config.Before, &printlogs, task.config.Command}
+	}
+	if context.contentSource == csapi.WorkspaceInitFromBackup {
+		// restart
+		return []*string{task.config.Before, task.config.Command}
+	}
+	// init
+	return []*string{task.config.Before, task.config.Init, task.config.Command}
+
+}
+
+func (task *task) prebuildLogFileName() string {
+	return "/workspace/.gitpod/prebuild-log-" + task.Id
+}
+
+func (tm *tasksManager) watch(task *task, terminal *terminal.Term) {
+	workspaceLog := log.WithField("component", "workspace")
+	stdout := terminal.Stdout.Listen()
+	start := time.Now()
+	prebuildChan := make(chan bool)
+	task.prebuildChan = &prebuildChan
+	go func() {
+		fileName := task.prebuildLogFileName()
+		file, err := os.Create(fileName)
+		if err != nil {
+			workspaceLog.WithError(err).Fatal("cannot create a prebuild log file")
+			prebuildChan <- false
+			return
+		}
+		defer file.Close()
+
+		fileWriter := bufio.NewWriter(file)
+
+		workspaceLog.Info("Writing build output to " + fileName)
+
+		buf := make([]byte, 4096)
+		for {
+			n, err := stdout.Read(buf)
+			if err == io.EOF {
+				elapsed := time.Since(start)
+				duration := ""
+				if elapsed >= 1*time.Minute {
+					elapsedInMinutes := strconv.Itoa(int(elapsed.Minutes()))
+					duration = "🎉 You just saved " + elapsedInMinutes + " minute"
+					if elapsedInMinutes != "1" {
+						duration += "s"
+					}
+					duration += " of watching your code build.\r\n"
+				}
+				data := string(buf[:n])
+				fileWriter.Write(buf[:n])
+				workspaceLog.WithField("type", "workspaceTaskOutput").WithField("data", data).Info()
+
+				endMessage := "\r\n🍌 This task ran as part of a workspace prebuild.\r\n" + duration + "\r\n"
+				fileWriter.WriteString(endMessage)
+				workspaceLog.WithField("type", "workspaceTaskOutput").WithField("data", endMessage).Info()
+
+				fileWriter.Flush()
+				prebuildChan <- true
+				break
+			}
+			if err != nil {
+				workspaceLog.WithError(err).Fatal("cannot read from a task terminal")
+				prebuildChan <- false
+				return
+			}
+			data := string(buf[:n])
+			fileWriter.Write(buf[:n])
+			workspaceLog.WithField("type", "workspaceTaskOutput").WithField("data", data).Info()
+		}
+	}()
+}
+
+func (tm *tasksManager) report(ctx context.Context) {
+	workspaceLog := log.WithField("component", "workspace")
+	ok := true
+	for _, task := range tm.tasks {
+		if task.prebuildChan != nil {
+			select {
+			case <-ctx.Done():
+				return
+			case prebuildOk := <-*task.prebuildChan:
+				if !prebuildOk {
+					ok = false
+				}
+			}
+		}
+	}
+	workspaceLog.WithField("type", "workspaceTaskOutput").WithField("data", "🚛 uploading prebuilt workspace").Info()
+	if ok {
```

Comment:
Nit: the more Go-ish way to write this would be
```go
if !ok {
	workspaceLog.WithField("type", "workspaceTaskFailed").WithField("error", "one of the tasks failed with non-zero exit code").Info()
	return
}

workspaceLog.WithField("type", "workspaceTaskDone").Info()
```